File: src/main/java/al132/alchemistry/blocks/fission/FissionControllerBlock.java
package al132.alchemistry.blocks.fission;

import al132.alchemistry.Config;
import al132.alchemistry.blocks.BaseTileBlock;
import al132.alchemistry.blocks.PowerStatus;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.material.Material;
import net.minecraft.client.resources.I18n;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.item.BlockItemUseContext;
import net.minecraft.item.ItemStack;
import net.minecraft.state.DirectionProperty;
import net.minecraft.state.EnumProperty;
import net.minecraft.state.StateContainer;
import net.minecraft.state.properties.BlockStateProperties;
import net.minecraft.util.Direction;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.StringTextComponent;
import net.minecraft.world.IBlockReader;

import javax.annotation.Nullable;
import java.util.List;

public class FissionControllerBlock extends BaseTileBlock {
    public static final EnumProperty<PowerStatus> STATUS = EnumProperty.create("status", PowerStatus.class, PowerStatus.values());
    public static final DirectionProperty FACING = BlockStateProperties.HORIZONTAL_FACING;


    public FissionControllerBlock() {
        super("fission_controller", FissionTile::new, Block.Properties.create(Material.IRON));
        this.setDefaultState(this.getDefaultState().with(STATUS, PowerStatus.OFF).with(FACING, Direction.NORTH));
    }

    @Override
    protected void fillStateContainer(StateContainer.Builder<Block, BlockState> builder) {
        builder.add(STATUS, FACING);
    }

    @Override
    @Nullable
    public BlockState getStateForPlacement(BlockItemUseContext context) {
        return this.getDefaultState()
                .with(FACING, context.getPlacementHorizontalFacing().getOpposite())
                .with(STATUS, PowerStatus.OFF);
    }

    @Override
    public void addInformation(ItemStack stack, @Nullable IBlockReader worldIn, List<ITextComponent> tooltip, ITooltipFlag flagIn) {
        super.addInformation(stack, worldIn, tooltip, flagIn);
        tooltip.add(new StringTextComponent(I18n.format("tooltip.alchemistry.energy_requirement", Config.FISSION_ENERGY_PER_TICK.get())));
    }


}

File: src/main/java/al132/alchemistry/blocks/FissionControllerBlock.kt
package al132.alchemistry.blocks

import al132.alchemistry.ConfigHandler
import al132.alchemistry.items.TooltipItemBlock
import al132.alib.utils.Translator
import al132.alib.utils.extensions.translate
import net.minecraft.block.properties.IProperty
import net.minecraft.block.properties.PropertyDirection
import net.minecraft.block.properties.PropertyEnum
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.Item
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.event.RegistryEvent

class FissionControllerBlock(name: String,
                             tileClass: Class<out TileEntity>,
                             guiID: Int)
    : BaseTileBlock(name, tileClass, guiID) {

    init {
        this.defaultState = this.blockState.baseState.withProperty(FACING, EnumFacing.NORTH)
                .withProperty(STATUS, PropertyPowerStatus.OFF)

    }

    override fun registerItemBlock(event: RegistryEvent.Register<Item>) {
        event.registry.register(TooltipItemBlock(this,
                Translator.translateToLocalFormatted("tooltip.alchemistry.energy_requirement",
                        ConfigHandler.fissionEnergyPerTick ?: "?"))
                .setRegistryName(this.registryName))
    }

    override fun createBlockState() = BlockStateContainer(this, *PROPERTIES)

    override fun getStateForPlacement(world: World, pos: BlockPos, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float, meta: Int, placer: EntityLivingBase, hand: EnumHand): IBlockState {
        val state = super.getStateForPlacement(world, pos, facing, hitX, hitY, hitZ, meta, placer, hand)
        return state.withProperty(FACING, placer.horizontalFacing.opposite).withProperty(STATUS, PropertyPowerStatus.OFF)
    }

    override fun getStateFromMeta(meta: Int): IBlockState {
        lateinit var facing: EnumFacing
        lateinit var status: PropertyPowerStatus
        when (meta) {
            in 0 until 3  -> facing = EnumFacing.NORTH
            in 3 until 6  -> facing = EnumFacing.SOUTH
            in 6 until 9  -> facing = EnumFacing.WEST
            in 9 until 12 -> facing = EnumFacing.EAST
            else          -> facing = EnumFacing.NORTH
        }
        when (meta % 3) {
            0 -> status = PropertyPowerStatus.OFF
            1 -> status = PropertyPowerStatus.STANDBY
            2 -> status = PropertyPowerStatus.ON
        }
        /* var enumfacing = EnumFacing.byIndex(meta)
         if (enumfacing.axis == EnumFacing.Axis.Y) {
             enumfacing = EnumFacing.NORTH
         }
         return this.defaultState.withProperty(FACING, enumfacing)
         */
        return this.defaultState.withProperty(FACING, facing).withProperty(STATUS, status)
    }

    override fun getMetaFromState(state: IBlockState): Int {
        //val dir: Int = (state.getValue(FACING) as EnumFacing).index
        var sum = 0
        when (state.getValue(FACING)) {
            EnumFacing.NORTH -> sum += 0
            EnumFacing.SOUTH -> sum += 3
            EnumFacing.WEST  -> sum += 6
            EnumFacing.EAST  -> sum += 9
        }
        when (state.getValue(STATUS)) {
            PropertyPowerStatus.OFF     -> sum += 0
            PropertyPowerStatus.STANDBY -> sum += 1
            PropertyPowerStatus.ON      -> sum += 2
            else                        -> sum += 0
        }
        return sum
    }

    companion object {
        val FACING = PropertyDirection.create("facing", EnumFacing.Plane.HORIZONTAL)
        val STATUS = PropertyEnum.create("status", PropertyPowerStatus::class.java)
        val PROPERTIES = arrayOf<IProperty<*>>(FACING, STATUS)
    }
}